## Урок 4.

In [1]:
import numpy as np
from scipy.linalg import lu

### Практическое задание, часть 1

__1.__ Решить систему уравнений методом Гаусса:

$$\begin{cases}
x_{1}+x_{2}-x_{3}-2x_{4}=0, \\
2x_{1}+x_{2}-x_{3}+x_{4}=-2, \\
x_{1}+x_{2}-3x_{3}+x_{4}=4.
\end{cases}$$

Система уравнений недоопределенна.

Преобразуем систему в расширенную матрицу:

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 1 & -1 & -2 \\ 
2 & 1 & -1 & 1 \\ 
1 & 1 & -3 & 1
\end{matrix}\right|
\begin{matrix}
0\\ 
-2\\
4
\end{matrix}
\end{pmatrix}.$$

Вычтем из второй строки первую, умноженную на 2, и из третьей строки первую:

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 1 & -1 & -2 \\ 
0 & -1 & 1 & 5 \\ 
0 & 0 & -2 & 3
\end{matrix}\right|
\begin{matrix}
0\\ 
-2\\
4
\end{matrix}
\end{pmatrix}.$$

Это один из особых случаев при использовании метода Гаусса. В полученном ступенчатом виде мы имеем ступеньки из двух чисел, а не из одного, как это требуется. В таком случае система имеет бесконечное количество решений.

Последнее уравнение содержит две неизвестных. В данном случае можно рассматривать в качестве свободного параметра $x_{4}=c$. Тогда, выражая остальные переменные через $c$, получим:

$$x_{4}=c,$$
/
$$-2x_{3}+3c=4 \; \Leftrightarrow \; x_{3}=-(4-3с)/2,$$
/
$$-x_{2} - (4-3с)/2 + 5c=-2 \; \Leftrightarrow \; x_{2}=13с/2.$$
/
$$x_1+13c/2+(4-3c)/2-2c=0 \; \Leftrightarrow \; x_{1}=-2-3c.$$

Подставляя произвольные числа вместо $c$, мы получим частное решение. Например, при $c=0$:

$$x_{1}=-2,$$
/
$$x_{2}=0,$$
/
$$x_{3}=-2,$$
/
$$x_{4}=0.$$

Параметр $c$ может принимать бесконечное количество значений, при которых уравнения в системе будут обращаться в тождества.

Подставив общее решение в исходную систему, также получим тождества:

$$\begin{cases}
-2-3c+13c/2+4/2-3c/2-2c=0,\\
-4-6c+13c/2+4/2-3c/2+c=-2,\\
-2-3c+13c/2+12/2-9c/2+c=4.
\end{cases}$$

__2.__ Проверить на совместность и выяснить, сколько решений будет иметь система линейных уравнений:

   а) $\begin{cases}
3x_{1}-x_{2}+x_{3}=4, \\
2x_{1}-5x_{2}-3x_{3}=-17, \\
x_{1}+x_{2}-x_{3}=0;
\end{cases}$

$$A=\begin{pmatrix}
3 & -1 & 1\\ 
2 & -5 & -3\\ 
1 & 1 & -1 
\end{pmatrix}$$
/
$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
3 & -1 & 1\\ 
2 & -5 & -3\\ 
1 & 1 & -1 
\end{matrix}\right|
\begin{matrix}
4\\ 
-17\\
0
\end{matrix}
\end{pmatrix}.$$

In [2]:
x = [3, -1, 1]
y = [2, -5, -3]
z = [1, 1, -1]

a = np.array([x, y, z])
r = np.linalg.matrix_rank(a)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 3


In [3]:
x1 = [3, -1, 1, 4]
y1 = [2, -5, -3, -17]
z1 = [1, 1, -1, 0]

a1 = np.array([x1, y1, z1])
r = np.linalg.matrix_rank(a1)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 3


$rankA=rank \tilde A=n$, где $n$ — число неизвестных, то система определена, т. е. имеет единственное решение

   б) $\begin{cases}
2x_{1}-4x_{2}+6x_{3}=1, \\
x_{1}-2x_{2}+3x_{3}=-2, \\
3x_{1}-6x_{2}+9x_{3}=5;
\end{cases}$

$$A=\begin{pmatrix}
2 & -4 & 6\\ 
1 & -2 & 3\\ 
3 & -6 & 9 
\end{pmatrix}$$
/
$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
2 & -4 & 6\\ 
1 & -2 & 3\\ 
3 & -6 & 9  
\end{matrix}\right|
\begin{matrix}
1\\ 
-2\\
5
\end{matrix}
\end{pmatrix}.$$

In [4]:
q = [2, -4, 6]
b = [1, -2, 3]
c = [3, -6, 9]

f = np.array([q, b, c])
r = np.linalg.matrix_rank(f)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 1


In [5]:
q1 = [2, -4, 6, 1]
b1 = [1, -2, 3, -2]
c1 = [3, -6, 9, 5]

f1 = np.array([q1, b1, c1])
r = np.linalg.matrix_rank(f1)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 2


$rankA<rank \tilde A$ - система несовместна

   в) $\begin{cases}
x_{1}+2x_{2}+5x_{3}=4, \\
3x_{1}+x_{2}-8x_{3}=-2. 
\end{cases}$

$$A=\begin{pmatrix}
1 & 2 & 5\\ 
3 & 1 & -8 
\end{pmatrix}$$
/
$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 5\\ 
3 & 1 & -8   
\end{matrix}\right|
\begin{matrix}
4\\ 
-2
\end{matrix}
\end{pmatrix}.$$

In [6]:
e = [1, 2, 5]
d = [3, 1, -8]

u = np.array([e, d])
r = np.linalg.matrix_rank(u)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 2


In [7]:
e1 = [1, 2, 5, 4]
d1 = [3, 1, -8, -2]

u1 = np.array([e, d])
r = np.linalg.matrix_rank(u1)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 2


$rankA=rank \tilde A<n$, то система имеет бесконечное количество решений

__3.__ Проверить на совместность и выяснить, сколько решений будет иметь система линейных уравнений, заданная расширенной матрицей

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 3 & -2 & 4\\ 
0 & 5 & 0 & 1\\ 
0 & 0 & 3 & 0\\ 
0 & 0 & 0 & 2
\end{matrix}\right|
\begin{matrix}
3\\ 
2\\
4\\
1
\end{matrix}
\end{pmatrix}.$$

In [8]:
g = [1, 3, -2, 4]
h = [0, 5, 0, 1]
k = [0, 0, 3, 0]
l = [0, 0, 0, 2]

n = np.array([g, h, k, l])
r = np.linalg.matrix_rank(n)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 4


In [9]:
g1 = [1, 3, -2, 4, 3]
h1 = [0, 5, 0, 1, 2]
k1 = [0, 0, 3, 0, 4]
l1 = [0, 0, 0, 2, 1]

n1 = np.array([g1, h1, k1, l1])
r = np.linalg.matrix_rank(n1)

print(f'Ранг матрицы: {r}')

Ранг матрицы: 4


$rankA=rank \tilde A=n$, где $n$ — число неизвестных, то система определена, т. е. имеет единственное решение

Это видно и без Python.

__4.__ Дана система линейных уравнений, заданная расширенной матрицей

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{matrix}\right|
\begin{matrix}
a\\ 
b\\
c
\end{matrix}
\end{pmatrix}.$$

Найти соотношение между параметрами $a$, $b$ и $c$, при которых система является несовместной.

Вычтем из второй строки первую, умноженную на 4, и из третьей строки первую, умноженную на 7:

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3 \\ 
0 & -3 & -6 \\ 
0 & -6 & -12
\end{matrix}\right|
\begin{matrix}
a\\ 
(b-4a)\\
(c-7a)
\end{matrix}
\end{pmatrix}.$$

И вычтем из третьей строки вторую умноженную на 2:

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3 \\ 
0 & -3 & -6 \\ 
0 & 0 & 0
\end{matrix}\right|
\begin{matrix}
a\\ 
(b-4a)\\
(c+a-2b)
\end{matrix}
\end{pmatrix}.$$


Если $(c+a-2b)\neq 0$, то $rank \tilde A=3$, в то время как $rank A=2$. Это означает, что система в этом случае несовместна.

### Практическое задание, часть 2

__1.__ Решить систему уравнений методом Крамера:

   а) $\begin{cases}
x_{1}-2x_{2}=1 \\
3x_{1}-4x_{2}=7
\end{cases}$

Найдем определитель матрицы коэффициентов:

$$detA=\begin{vmatrix}
1 & -2\\ 
3 & -4
\end{vmatrix}=1 \cdot (-4)-3\cdot (-2)=2\neq 0,$$

следовательно, система совместна.

Найдем определители $detA_{1}$, $detA_{2}$:

$$detA_{1}=\begin{vmatrix}
1 & -2\\ 
7 & -4
\end{vmatrix}=1 \cdot (-4)-7\cdot (-2)=10$$
/
$$detA_{2}=\begin{vmatrix}
1 & 1\\ 
3 & 7
\end{vmatrix}=1 \cdot 7-3\cdot 1=4$$

Найдем решение по формулам Крамера:

$$x_{1} = \frac{detA_{1}}{detA} = \frac{10}{2}=5,$$
/
$$x_{2} = \frac{detA_{2}}{detA} = \frac{4}{2}=2$$

   б) $\begin{cases}
2x_{1}-x_{2}+5x_{3}=10 \\
x_{1}+x_{2}-3x_{3}=-2 \\
2x_{1}+4x_{2}+x_{3}=1
\end{cases}$

Найдем определитель матрицы коэффициентов:

$$detA=\begin{vmatrix}
2 & -1 & 5\\ 
1 & 1 & -3\\ 
2 & 4 & 1
\end{vmatrix}=
2\begin{vmatrix}
1 & -3\\ 
4 & 1 
\end{vmatrix}+
\begin{vmatrix}
1 & -3\\ 
2 & 1 
\end{vmatrix}+
5\begin{vmatrix}
1 & 1 \\ 
2 & 4
\end{vmatrix}=2(1 \cdot 1+3\cdot 4)+(1\cdot 1+2\cdot 3)+5(1\cdot 4-2\cdot 1)=43\neq 0,$$

следовательно, система совместна.

Найдем определители $detA_{1}$, $detA_{2}$, $detA_{3}$:

$$detA_{1}=\begin{vmatrix}
10 & -1 & 5\\ 
-2 & 1 & -3\\ 
1 & 4 & 1
\end{vmatrix}=
10\begin{vmatrix}
1 & -3\\ 
4 & 1 
\end{vmatrix}+
\begin{vmatrix}
-2 & -3\\ 
1 & 1 
\end{vmatrix}+
5\begin{vmatrix}
-2 & 1 \\ 
1 & 4
\end{vmatrix}=10(1 \cdot 1+3\cdot 4)+((-2)\cdot 1+1\cdot 3)+5(-2\cdot 4-1\cdot 1)=96,$$

$$detA_{2}=\begin{vmatrix}
2 & 10 & 5\\ 
1 & -2 & -3\\ 
2 & 1 & 1
\end{vmatrix}=
2\begin{vmatrix}
-2 & -3\\ 
1 & 1 
\end{vmatrix}-
10\begin{vmatrix}
1 & -3\\ 
2 & 1 
\end{vmatrix}+
5\begin{vmatrix}
1 & -2 \\ 
2 & 1
\end{vmatrix}=2(5 \cdot 1-0\cdot 3)-4(1\cdot 1-3\cdot 3)-1(1\cdot 0-3\cdot 5)=-43,$$

$$detA_{3}=\begin{vmatrix}
2 & -1 & 10\\ 
1 & 1 & -2\\ 
2 & 4 & 1
\end{vmatrix}=
2\begin{vmatrix}
1 & -2\\ 
4 & 1 
\end{vmatrix}+
\begin{vmatrix}
1 & -2\\ 
2 & 1 
\end{vmatrix}+
10\begin{vmatrix}
1 & 1 \\ 
2 & 4
\end{vmatrix}=2(1 \cdot 1-4\cdot (-2))+(1\cdot 1-2\cdot (-2))+10(1\cdot 4-2\cdot 1)=43.$$

Найдем решение по формулам Крамера:

$$x_{1} = \frac{detA_{1}}{detA} = \frac{96}{43}=2,$$
/
$$x_{2} = \frac{detA_{2}}{detA} = \frac{-43}{43}=-1,$$
/
$$x_{3} = \frac{detA_{3}}{detA} = \frac{43}{43}=1.$$

__2*.__ Найти $L$-матрицу $LU$-разложения для матрицы коэффициентов:

   а)$$\begin{pmatrix}
1 & 2 & 4 \\ 
2 & 9 & 12 \\ 
3 & 26 & 30
\end{pmatrix}$$
/
$$U=\begin{pmatrix}
1 & 2 & 4 \\ 
0 & 5 & 4 \\ 
0 & 0 & 2
\end{pmatrix},$$
/
$$L=\begin{pmatrix}
1 & 0 & 0 \\ 
2 & 1 & 0 \\ 
3 & 4 & 1
\end{pmatrix}.$$
    
   б)$$\begin{pmatrix}
1 & 1 & 2 & 4\\ 
2 & 5 & 8 & 9\\ 
3 & 18 & 29 & 18\\
4 & 22 & 53 & 33
\end{pmatrix}$$
/
$$U=\begin{pmatrix}
1 & 1 & 2 & 4 \\ 
0 & 3 & 4 & 1 \\
0 & 0 & 3 & 1 \\
0 & 0 & 0 & 4
\end{pmatrix},$$
/
$$L=\begin{pmatrix}
1 & 0 & 0 & 0 \\ 
2 & 1 & 0 & 0 \\
3 & 5 & 1 & 0 \\
4 & 6 & 7 & 1
\end{pmatrix}.$$

__3*.__ Решить систему линейных уравнений методом $LU$-разложения

$$\begin{cases}
2x_{1}+x_{2}+3x_{3}=1 \\
11x_{1}+7x_{2}+5x_{3}=-6 \\
9x_{1}+8x_{2}+4x_{3}=-5
\end{cases}$$

In [12]:
A = np.array([[2, 1, 3], [11, 7, 5], [9, 8, 4]])
p, l, u = lu(A)
print(f'L-матрица:\n{l}')
print(f'U-матрица:\n{u}')
print(f'Перестановка строк:{p}')

L-матрица:
[[ 1.          0.          0.        ]
 [ 0.81818182  1.          0.        ]
 [ 0.18181818 -0.12        1.        ]]
U-матрица:
[[11.          7.          5.        ]
 [ 0.          2.27272727 -0.09090909]
 [ 0.          0.          2.08      ]]
Перестановка строк:[[0. 0. 1.]
 [1. 0. 0.]
 [0. 1. 0.]]


$$Ly=b:$$

$$\begin{cases}
y_{1}=1, \\
0.82 y_{1}+y_{2}=-6, \\
0.18 y_{1}-0.12y_{2}+y_{3}=-5.
\end{cases}$$

$$y_{1}=1,$$
$$y_{2}=-6.82,$$
$$y_{3}=-6.$$

И затем систему

$$Ux=y:$$

$$\begin{cases}
11x_{1}+7x_{2}+5x_{3}=1, \\
2.27x_{2}-0.09x_3=-6.82, \\
2.08x_{3}=-6.
\end{cases}$$

$$x_{3}=-2.8838,$$
/
$$x_{2}=-3.1187,$$
/
$$x_{1}=3.3864.$$

__4*.__ Решить систему линейных уравнений методом Холецкого

$$\begin{cases}
81x_{1}-45x_{2}+45x_{3}=531 \\
-45x_{1}+50x_{2}-15x_{3}=-460 \\
45x_{1}-15x_{2}+38x_{3}=193
\end{cases}$$

__5*.__ Написать на Python программу с реализацией одного из изученных алгоритмов решения СЛАУ.